src/core/Worker.hpp

```diff
@@ -87,59 +102,72 @@ class Worker
             pipeline_.SubscribeTo(endpoint);
         }
     }
-    auto stop_worker() noexcept -> std::shared_future<void>
-    {
-        pipeline_.Close();
-        downcast().shutdown(shutdown_promise_);
-
-        return shutdown_;
-    }
 
     Worker(
         const API& api,
         const std::chrono::milliseconds rateLimit,
         const Vector<network::zeromq::SocketData>& extra = {}) noexcept
-        : api_(api)
-        , rate_limit_(rateLimit)
-        , running_(true)
-        , shutdown_promise_()
-        , shutdown_(shutdown_promise_.get_future())
-        , pipeline_(api.Network().ZeroMQ().Internal().Pipeline(
-              [this](auto&& in) { downcast().pipeline(std::move(in)); },
+        : api_{api}
+        , rate_limit_{rateLimit}
+        , running_{true}
+        , shutdown_once_flag_{}
+        , shutdown_promise_{}
+        , shutdown_complete_{shutdown_promise_.get_future()}
+        , last_executed_{Clock::now()}
+        , state_machine_queued_{false}
+        , pipeline_{api.Network().ZeroMQ().Internal().Pipeline(
+              [this](auto&& in) { pipeline(std::move(in)); },
               {},
               {},
               {},
-              extra))
-        , last_executed_(Clock::now())
-        , state_machine_queued_(false)
+              extra)}
     {
         LogTrace()(OT_PRETTY_CLASS())("using ZMQ batch ")(pipeline_.BatchID())
             .Flush();
     }
 
-    ~Worker() { stop_worker().get(); }
+    virtual ~Worker()
+    {
+        if (check_running_mark_stopped()) {
+            close_pipeline();
+            mark_shut_down();
```

Comment:
Even though `mark_shut_down` is marked `noexcept`, `std::call_once` used there can still throw `std::system_error` if any condition prevents calls to `std::call_once` from executing as specified. And as we know throwing in dectructors is not a good idea. There are a lot of places where `mark_shut_down` is called in destructor.

Reply:
You are right. The whole business of how noexcept qualifiers and const qualifiers are used in this code is a great source of concern. We'll need to address the issue as part of the refactor.